From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Sudzzy <originmc@outlook.com>
Date: Wed, 2 Mar 2016 23:34:44 -0600
Subject: [PATCH] Configurable container update tick rate


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 8f45fa5ca878428a9ce6b18df075d3f09b2cf757..1d8d1a4e387f205f6c1c9b608f20e30763dec01a 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -177,4 +177,9 @@ public class PaperWorldConfig {
     private void mobSpawnerTickRate() {
         mobSpawnerTickRate = getInt("mob-spawner-tick-rate", 1);
     }
+
+    public int containerUpdateTickRate;
+    private void containerUpdateTickRate() {
+        containerUpdateTickRate = getInt("container-update-tick-rate", 1);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 474f727ad631468251578bf847319d1c7e0b0f19..92885ea93f0d33ec68bca2ae6f6e83450d84a76b 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -205,6 +205,7 @@ public class ServerPlayer extends Player implements ContainerListener {
     public boolean ignoreSlotUpdateHack;
     public int latency;
     public boolean wonGame;
+    private int containerUpdateDelay; // Paper
 
     // CraftBukkit start
     public String displayName;
@@ -527,7 +528,12 @@ public class ServerPlayer extends Player implements ContainerListener {
             --this.invulnerableTime;
         }
 
-        this.containerMenu.broadcastChanges();
+        // Paper start - Configurable container update tick rate
+        if (--containerUpdateDelay <= 0) {
+            this.containerMenu.broadcastChanges();
+            containerUpdateDelay = level.paperConfig.containerUpdateTickRate;
+        }
+        // Paper end
         if (!this.level.isClientSide && !this.containerMenu.stillValid(this)) {
             this.closeContainer();
             this.containerMenu = this.inventoryMenu;
