From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Sat, 6 Oct 2018 21:47:44 -0500
Subject: [PATCH] Allow setting the vex's summoner


diff --git a/src/main/java/net/minecraft/world/entity/monster/Vex.java b/src/main/java/net/minecraft/world/entity/monster/Vex.java
index b5730441895afafaf6df48d6e2bd90ccfe9aa90b..2048e029a0989bc0db0bd6789be64156ab9d1778 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Vex.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Vex.java
@@ -166,6 +166,7 @@ public class Vex extends Monster {
         this.setVexFlag(1, charging);
     }
 
+    public void setOwner(Mob entityinsentient) { setOwner(entityinsentient); } // Paper - OBFHELPER
     public void setOwner(Mob owner) {
         this.owner = owner;
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftVex.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftVex.java
index 4a98770b7b511d8b4bfd379f88242245858ad858..6b55293cd7bb310400034007e082d531a2e15016 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftVex.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftVex.java
@@ -21,6 +21,10 @@ public class CraftVex extends CraftMonster implements Vex {
         net.minecraft.world.entity.Mob owner = getHandle().getOwner();
         return owner != null ? (Mob) owner.getBukkitEntity() : null;
     }
+
+    public void setSummoner(Mob summoner) {
+        getHandle().setOwner(summoner == null ? null : ((CraftMob) summoner).getHandle());
+    }
     // Paper end
 
     @Override
