From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GreenMeanie <GreenMeanieMC@gmail.com>
Date: Sat, 20 Oct 2018 22:34:02 -0400
Subject: [PATCH] Reset players airTicks on respawn


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index fb627ef1ceff7c35c34219fd239611a304f7e55b..8c2aa4a8d2224cf7b2b201a2ccab08cf85d8df73 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -2130,6 +2130,7 @@ public class ServerPlayer extends Player implements ContainerListener {
         }
 
         this.setHealth(this.getMaxHealth());
+        this.setAirSupply(this.getMaxAirTicks()); // Paper
         this.remainingFireTicks = 0;
         this.fallDistance = 0;
         this.foodData = new FoodData(this);
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index ac0bda591b9f6c327b91daefc997d9c947c17441..9ba8618a62b52695215fe622226e873d2306087a 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -2352,6 +2352,7 @@ public abstract class Entity implements Nameable, CommandSource, KeyedObject { /
 
     }
 
+    public final int getMaxAirTicks() { return getMaxAirSupply(); } // Paper - OBFHELPER
     public int getMaxAirSupply() {
         return 300;
     }
