From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Mon, 27 May 2019 17:35:39 -0500
Subject: [PATCH] MC-114618 - Fix EntityAreaEffectCloud from going negative
 size


diff --git a/src/main/java/net/minecraft/world/entity/AreaEffectCloud.java b/src/main/java/net/minecraft/world/entity/AreaEffectCloud.java
index 842e80b176f66d23e576c8f50f2233f8502bd5dc..65866f74697f0496f259ef84f3bc726f5b2d2795 100644
--- a/src/main/java/net/minecraft/world/entity/AreaEffectCloud.java
+++ b/src/main/java/net/minecraft/world/entity/AreaEffectCloud.java
@@ -195,6 +195,12 @@ public class AreaEffectCloud extends Entity {
         super.tick();
         boolean flag = this.isWaiting();
         float f = this.getRadius();
+        // Paper start - fix MC-114618
+        if (f < 0.0F) {
+            this.remove();
+            return;
+        }
+        // Paper end
 
         if (this.level.isClientSide) {
             ParticleOptions particleparam = this.getParticle();
