From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <blake.galbreath@gmail.com>
Date: Sat, 13 Apr 2019 16:50:58 -0500
Subject: [PATCH] Add option to allow iron golems to spawn in air


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 89f6de173225315783a09aae2fe77a3ede2b78d5..fd0e58574f4f07913161be489d031a011f059f8e 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -381,6 +381,11 @@ public class PaperWorldConfig {
         scanForLegacyEnderDragon = getBoolean("game-mechanics.scan-for-legacy-ender-dragon", true);
     }
 
+    public boolean ironGolemsCanSpawnInAir = false;
+    private void ironGolemsCanSpawnInAir() {
+        ironGolemsCanSpawnInAir = getBoolean("iron-golems-can-spawn-in-air", ironGolemsCanSpawnInAir);
+    }
+
     public boolean armorStandEntityLookups = true;
     private void armorStandEntityLookups() {
         armorStandEntityLookups = getBoolean("armor-stands-do-collision-entity-lookups", true);
diff --git a/src/main/java/net/minecraft/world/entity/animal/IronGolem.java b/src/main/java/net/minecraft/world/entity/animal/IronGolem.java
index 810b50d6f4e6b07dda3b0b2629fea428f624b107..584bd3ea3638583b6f11458650f03f8915d92161 100644
--- a/src/main/java/net/minecraft/world/entity/animal/IronGolem.java
+++ b/src/main/java/net/minecraft/world/entity/animal/IronGolem.java
@@ -297,7 +297,7 @@ public class IronGolem extends AbstractGolem implements NeutralMob {
         BlockPos blockposition1 = blockposition.below();
         BlockState iblockdata = iworldreader.getBlockState(blockposition1);
 
-        if (!iblockdata.entityCanStandOn((BlockGetter) iworldreader, blockposition1, (Entity) this)) {
+        if (!iblockdata.entityCanStandOn((BlockGetter) iworldreader, blockposition1, (Entity) this) && !level.paperConfig.ironGolemsCanSpawnInAir) { // Paper
             return false;
         } else {
             for (int i = 1; i < 3; ++i) {
