From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AvrooVulcan <avrovulcan.programming@gmail.com>
Date: Fri, 17 Apr 2020 00:15:23 +0100
Subject: [PATCH] Broadcast join message to console


diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 5cfa0956d4b70f9d37d638bdf8a9dbe7e813b748..71bdd933b5970c4053228d23bbb76b6ab90dac55 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -285,9 +285,9 @@ public abstract class PlayerList {
         joinMessage = playerJoinEvent.getJoinMessage();
 
         if (joinMessage != null && joinMessage.length() > 0) {
-            for (Component line : org.bukkit.craftbukkit.util.CraftChatMessage.fromString(joinMessage)) {
-                server.getPlayerList().broadcastAll(new ClientboundChatPacket(line, ChatType.SYSTEM, Util.NIL_UUID));
-            }
+            // Paper start - Removed sendAll for loop and broadcasted to console also
+            server.getPlayerList().sendMessage(CraftChatMessage.fromString(joinMessage));
+            // Paper end
         }
         // CraftBukkit end
 
