From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mariell Hoversholm <proximyst@proximyst.com>
Date: Fri, 14 Aug 2020 23:41:19 +0200
Subject: [PATCH] Don't mark null chunk sections for block updates


diff --git a/src/main/java/net/minecraft/server/level/ChunkHolder.java b/src/main/java/net/minecraft/server/level/ChunkHolder.java
index 361df5329e184b4ce2b5ceb0bdc12f613d59c30e..6e372302eff34e878a37ed6d257ceb7317997b5e 100644
--- a/src/main/java/net/minecraft/server/level/ChunkHolder.java
+++ b/src/main/java/net/minecraft/server/level/ChunkHolder.java
@@ -443,6 +443,7 @@ public class ChunkHolder {
                         this.broadcastBlockEntityIfNeeded(world, blockposition, iblockdata);
                     } else {
                         LevelChunkSection chunksection = chunk.getSections()[sectionposition.getY()];
+                        if (chunksection == null) chunksection = new LevelChunkSection(sectionposition.getY(), chunk, world, true); // Paper - make a new chunk section if none was found
                         ClientboundSectionBlocksUpdatePacket packetplayoutmultiblockchange = new ClientboundSectionBlocksUpdatePacket(sectionposition, shortset, chunksection, this.resendLight);
 
                         this.broadcast(packetplayoutmultiblockchange, false);
