From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: foss-mc <69294560+foss-mc@users.noreply.github.com>
Date: Sun, 30 Aug 2020 15:30:29 +0800
Subject: [PATCH] Fix CraftTeam null check


diff --git a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java
index 6811485c27697be33549c6735011298a3fb6bc89..6554c0539c858a15d8e1b5c5056349e81b643463 100644
--- a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java
+++ b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java
@@ -202,7 +202,7 @@ final class CraftTeam extends CraftScoreboardComponent implements Team {
 
     @Override
     public boolean hasEntry(String entry) throws IllegalArgumentException, IllegalStateException {
-        Validate.notNull("Entry cannot be null");
+        Validate.notNull(entry, "Entry cannot be null"); // Paper
 
         CraftScoreboard scoreboard = checkState();
 
