From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mariell Hoversholm <proximyst@proximyst.com>
Date: Wed, 30 Sep 2020 22:49:14 +0200
Subject: [PATCH] Toggle for removing existing dragon


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 8923314d731dea5c2707dbb271b8b0819d826a2e..9a0ade5875c34487a65f82f9380f9d25b4432586 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -672,4 +672,12 @@ public class PaperWorldConfig {
             log("Using vanilla redstone algorithm.");
         }
     }
+
+    public boolean shouldRemoveDragon = false;
+    private void shouldRemoveDragon() {
+        shouldRemoveDragon = getBoolean("should-remove-dragon", shouldRemoveDragon);
+        if (shouldRemoveDragon) {
+            log("The Ender Dragon will be removed if she already exists without a portal.");
+        }
+    }
 }
diff --git a/src/main/java/net/minecraft/world/level/dimension/end/EndDragonFight.java b/src/main/java/net/minecraft/world/level/dimension/end/EndDragonFight.java
index 0131e98e13ecec4efc43e62f4d4ee45380c7979f..9a878925fd1b042655afa1150db4f63b3a1f842b 100644
--- a/src/main/java/net/minecraft/world/level/dimension/end/EndDragonFight.java
+++ b/src/main/java/net/minecraft/world/level/dimension/end/EndDragonFight.java
@@ -214,7 +214,7 @@ public class EndDragonFight {
             this.dragonUUID = entityenderdragon.getUUID();
             EndDragonFight.LOGGER.info("Found that there's a dragon still alive ({})", entityenderdragon);
             this.dragonKilled = false;
-            if (!flag) {
+            if (!flag && this.level.paperConfig.shouldRemoveDragon) { // Paper
                 EndDragonFight.LOGGER.info("But we didn't have a portal, let's remove it.");
                 entityenderdragon.remove();
                 this.dragonUUID = null;
