From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Mon, 29 Feb 2016 20:40:33 -0600
Subject: [PATCH] Paper build config changes


diff --git a/build.gradle.kts b/build.gradle.kts
index 150c69b8f5e6d3bb51b855efe7b433e7a2cdf3e6..95652a9187952820a23506e8ebf4f40af44161e8 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -15,6 +15,8 @@ repositories {
 
 dependencies {
     implementation(project(":Paper-API"))
+    implementation(project(":Paper-MojangAPI"))
+
     implementation("jline:jline:2.12.1")
     implementation("org.apache.logging.log4j:log4j-iostreams:2.8.1")
     implementation("org.ow2.asm:asm:8.0.1")
@@ -22,8 +24,13 @@ dependencies {
         // This includes junit transitively for whatever reason
         isTransitive = false
     }
+    // wrapper to use either java 8 sun cleaner or java9+ official cleaner
+    implementation("co.aikar:cleaner:1.0-SNAPSHOT")
+    implementation("io.netty:netty-all:4.1.50.Final")
+
     runtimeOnly("org.xerial:sqlite-jdbc:3.32.3")
     runtimeOnly("mysql:mysql-connector-java:5.1.49")
+
     testImplementation("junit:junit:4.13.1")
     testImplementation("org.hamcrest:hamcrest-library:1.3")
 }
@@ -48,7 +55,7 @@ tasks.jar {
 
 tasks.shadowJar {
     listOf(
-        "jline", "it.unimi", "org.apache.commons.codec", "org.apache.commons.io",
+        "jline", "org.apache.commons.codec", "org.apache.commons.io",
         "org.apache.commons.lang3", "org.objectweb.asm"
     ).forEach { pack ->
         relocate(pack, "org.bukkit.craftbukkit.libs.$pack")
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 60ede7442b2d9a63fcd765aff29a96c9130aebac..3f52c19b6b797e26b3dcedcec2afd28c478e7091 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -186,7 +186,7 @@ public class Main {
                 }
 
                 if (Main.class.getPackage().getImplementationVendor() != null && System.getProperty("IReallyKnowWhatIAmDoingISwear") == null) {
-                    Date buildDate = new Date(Integer.parseInt(Main.class.getPackage().getImplementationVendor()) * 1000L);
+                    Date buildDate = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'").parse(Main.class.getPackage().getImplementationVendor()); // Paper
 
                     Calendar deadline = Calendar.getInstance();
                     deadline.add(Calendar.DAY_OF_YEAR, -28);
diff --git a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
index 93046379d0cefd5d3236fc59e698809acdc18f80..674096cab190d62622f9947853b056f57d43a2a5 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
@@ -11,7 +11,7 @@ public final class Versioning {
     public static String getBukkitVersion() {
         String result = "Unknown-Version";
 
-        InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("META-INF/maven/org.spigotmc/spigot-api/pom.properties");
+        InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("META-INF/maven/com.destroystokyo.paper/paper-api/pom.properties");
         Properties properties = new Properties();
 
         if (stream != null) {
