From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Byteflux <byte@byteflux.net>
Date: Tue, 1 Mar 2016 15:08:03 -0600
Subject: [PATCH] Remove invalid mob spawner tile entities


diff --git a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
index dd501797bb2a52e2434bb868e039327bf0b3a866..440481a6aef5169d2e477cb1bc8ddbd15059ac77 100644
--- a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
@@ -44,7 +44,9 @@ import net.minecraft.world.level.block.BaseEntityBlock;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.EntityBlock;
+import net.minecraft.world.level.block.SpawnerBlock;
 import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.level.block.entity.SpawnerBlockEntity;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.levelgen.DebugLevelSource;
 import net.minecraft.world.level.levelgen.Heightmap;
@@ -640,6 +642,10 @@ public class LevelChunk implements ChunkAccess {
             }
 
             // CraftBukkit start
+            // Paper start - Remove invalid mob spawner tile entities
+        } else if (blockEntity instanceof SpawnerBlockEntity && !(getBlockData(pos.getX(), pos.getY(), pos.getZ()).getBlock() instanceof SpawnerBlock)) {
+            this.blockEntities.remove(pos);
+            // Paper end
         } else {
             System.out.println("Attempted to place a tile entity (" + blockEntity + ") at " + blockEntity.worldPosition.getX() + "," + blockEntity.worldPosition.getY() + "," + blockEntity.worldPosition.getZ()
                 + " (" + getBlockState(pos) + ") where there was no entity tile!");
