From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 8 Mar 2015 22:55:25 -0600
Subject: [PATCH] Optimize TileEntity Ticking


diff --git a/src/main/java/co/aikar/timings/TimingsExport.java b/src/main/java/co/aikar/timings/TimingsExport.java
index a50a3c22d314f06f8aaff25da194a5e118393569..f7fc080089f5c54b0b7cdc3baa8ed18f493ead32 100644
--- a/src/main/java/co/aikar/timings/TimingsExport.java
+++ b/src/main/java/co/aikar/timings/TimingsExport.java
@@ -109,7 +109,7 @@ public class TimingsExport extends Thread {
             pair("end", System.currentTimeMillis() / 1000),
             pair("online-mode", Bukkit.getServer().getOnlineMode()),
             pair("sampletime", (System.currentTimeMillis() - TimingsManager.timingStart) / 1000),
-            pair("datapacks", toArrayMapper(MinecraftServer.getServer().getPackRepository().getSelectedIds(), pack -> {
+            pair("datapacks", toArrayMapper(MinecraftServer.getServer().getPackRepository().getSelectedPacks(), pack -> {
                 // Don't feel like obf helper'ing these, non fatal if its temp missed.
                 return ChatColor.stripColor(CraftChatMessage.fromComponent(pack.getChatLink(true)));
             }))
@@ -148,8 +148,8 @@ public class TimingsExport extends Thread {
         );
 
         parent.put("worlds", toObjectMapper(MinecraftServer.getServer().getAllLevels(), world -> {
-            if (world.getWorldData().getName().equals("worldeditregentempworld")) return null;
-            return pair(world.getWorldData().getName(), createObject(
+            if (world.getWorld().getName().equals("worldeditregentempworld")) return null;
+            return pair(world.getWorld().getName(), createObject(
                 pair("gamerules", toObjectMapper(world.getWorld().getGameRules(), rule -> {
                     return pair(rule, world.getWorld().getGameRuleValue(rule));
                 })),
diff --git a/src/main/java/net/minecraft/world/level/block/ChestBlock.java b/src/main/java/net/minecraft/world/level/block/ChestBlock.java
index 3447e935469ee95444fb25a1ebc2c805e6229250..0cefc9c12aa58f61607fd9d825568aba5df95772 100644
--- a/src/main/java/net/minecraft/world/level/block/ChestBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/ChestBlock.java
@@ -54,8 +54,8 @@ import net.minecraft.world.phys.shapes.VoxelShape;
 public class ChestBlock extends AbstractChestBlock<ChestBlockEntity> implements SimpleWaterloggedBlock {
 
     public static final DirectionProperty FACING = HorizontalDirectionalBlock.FACING;
-    public static final EnumProperty<ChestType> TYPE = BlockStateProperties.CHEST_TYPE;
-    public static final BooleanProperty WATERLOGGED = BlockStateProperties.WATERLOGGED;
+    public static final EnumProperty<ChestType> TYPE = BlockStateProperties.CHEST_TYPE; public static final EnumProperty<ChestType> CHEST_TYPE_PROPERTY = TYPE; // Paper - OBFHELPER
+    public static final BooleanProperty WATERLOGGED = BlockStateProperties.WATERLOGGED; public static final BooleanProperty waterlogged() { return WATERLOGGED; } // Paper OBFHELPER
     protected static final VoxelShape NORTH_AABB = Block.box(1.0D, 0.0D, 0.0D, 15.0D, 14.0D, 15.0D);
     protected static final VoxelShape SOUTH_AABB = Block.box(1.0D, 0.0D, 1.0D, 15.0D, 14.0D, 16.0D);
     protected static final VoxelShape WEST_AABB = Block.box(0.0D, 0.0D, 1.0D, 15.0D, 14.0D, 15.0D);
diff --git a/src/main/java/net/minecraft/world/level/block/entity/ChestBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/ChestBlockEntity.java
index ac4635b76a9d09eba142af464d388373d33f4e02..260cf6ddc77638bbfda3438ed4066be95dd20437 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/ChestBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/ChestBlockEntity.java
@@ -8,6 +8,7 @@ import net.minecraft.core.NonNullList;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.TranslatableComponent;
+import net.minecraft.server.MCUtil;
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.sounds.SoundSource;
@@ -32,7 +33,7 @@ import org.bukkit.craftbukkit.entity.CraftHumanEntity;
 import org.bukkit.entity.HumanEntity;
 // CraftBukkit end
 
-public class ChestBlockEntity extends RandomizableContainerBlockEntity implements TickableBlockEntity {
+public class ChestBlockEntity extends RandomizableContainerBlockEntity { // Paper - Remove ITickable
 
     private NonNullList<ItemStack> items;
     protected float openness;
@@ -110,14 +111,20 @@ public class ChestBlockEntity extends RandomizableContainerBlockEntity implement
         return tag;
     }
 
-    @Override
     public void tick() {
         int i = this.worldPosition.getX();
         int j = this.worldPosition.getY();
         int k = this.worldPosition.getZ();
 
         ++this.tickInterval;
-        this.openCount = getOpenCount(this.level, this, this.tickInterval, i, j, k, this.openCount);
+    }
+
+    public void doOpenLogic() {
+        int i = this.worldPosition.getX();
+        int j = this.worldPosition.getY();
+        int k = this.worldPosition.getZ();
+
+        //this.viewingCount = a(this.world, this, this.j, i, j, k, this.viewingCount); // Paper - check is faulty given our logic is called before active container set
         this.oOpenness = this.openness;
         float f = 0.1F;
 
@@ -131,25 +138,31 @@ public class ChestBlockEntity extends RandomizableContainerBlockEntity implement
         if (this.openCount > 0 && this.openness == 0.0F) {
             this.playSound(SoundEvents.CHEST_OPEN);
         }
+    }
 
-        if (this.openCount == 0 && this.openness > 0.0F || this.openCount > 0 && this.openness < 1.0F) {
-            float f1 = this.openness;
+    public void doCloseLogic() {
+        if (this.openCount == 0 /* && this.a > 0.0F || this.viewingCount > 0 && this.a < 1.0F */) { // Paper - disable all but player count check
+            /* // Paper - disable animation stuff
+            float f1 = this.a;
 
-            if (this.openCount > 0) {
-                this.openness += 0.1F;
+            if (this.viewingCount > 0) {
+                this.a += 0.1F;
             } else {
-                this.openness -= 0.1F;
+                this.a -= 0.1F;
             }
 
-            if (this.openness > 1.0F) {
-                this.openness = 1.0F;
+            if (this.a > 1.0F) {
+                this.a = 1.0F;
             }
 
             float f2 = 0.5F;
 
-            if (this.openness < 0.5F && f1 >= 0.5F) {
+            if (this.a < 0.5F && f1 >= 0.5F) {
+            */
+            MCUtil.scheduleTask(10, () -> {
                 this.playSound(SoundEvents.CHEST_CLOSE);
-            }
+                }, "Chest Sounds");
+            //} // Paper end
 
             if (this.openness < 0.0F) {
                 this.openness = 0.0F;
@@ -188,6 +201,7 @@ public class ChestBlockEntity extends RandomizableContainerBlockEntity implement
     }
 
     public void playSound(SoundEvent soundeffect) {
+        if (!this.getBlockState().contains(ChestBlock.CHEST_TYPE_PROPERTY)) { return; } // Paper - this can be delayed, double check exists - Fixes GH-2074
         ChestType blockpropertychesttype = (ChestType) this.getBlockState().getValue(ChestBlock.TYPE);
 
         if (blockpropertychesttype != ChestType.LEFT) {
@@ -226,6 +240,7 @@ public class ChestBlockEntity extends RandomizableContainerBlockEntity implement
 
             ++this.openCount;
             if (this.level == null) return; // CraftBukkit
+            doOpenLogic(); // Paper
 
             // CraftBukkit start - Call redstone event
             if (this.getBlockState().getBlock() == Blocks.TRAPPED_CHEST) {
@@ -248,6 +263,7 @@ public class ChestBlockEntity extends RandomizableContainerBlockEntity implement
             --this.openCount;
 
             // CraftBukkit start - Call redstone event
+            doCloseLogic(); // Paper
             if (this.getBlockState().getBlock() == Blocks.TRAPPED_CHEST) {
                 int newPower = Math.max(0, Math.min(15, this.openCount));
 
diff --git a/src/main/java/net/minecraft/world/level/block/entity/EnderChestBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/EnderChestBlockEntity.java
index 85f7c82fbce6209ef5ef151b096e371c6a4672a1..59f270f28cc08142e220ead8010a30a5e027a3f4 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/EnderChestBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/EnderChestBlockEntity.java
@@ -1,11 +1,12 @@
 package net.minecraft.world.level.block.entity;
 
+import net.minecraft.server.MCUtil;
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.sounds.SoundSource;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.block.Blocks;
 
-public class EnderChestBlockEntity extends BlockEntity implements TickableBlockEntity {
+public class EnderChestBlockEntity extends BlockEntity { // Paper - Remove ITickable
     public float openness;
     public float oOpenness;
     public int openCount;
@@ -15,41 +16,69 @@ public class EnderChestBlockEntity extends BlockEntity implements TickableBlockE
         super(BlockEntityType.ENDER_CHEST);
     }
 
-    @Override
+    // @Override // Paper - Remove ITickable
     public void tick() {
         if (++this.tickInterval % 20 * 4 == 0) {
             this.level.blockEvent(this.worldPosition, Blocks.ENDER_CHEST, 1, this.openCount);
         }
 
         this.oOpenness = this.openness;
+        /* // Paper
+        int i = this.position.getX();
+        int j = this.position.getY();
+        int k = this.position.getZ();
+        float f = 0.1F;
+        double d0;
+        // Paper start
+        */
+    }
+
+    private void doOpenLogic() {
         int i = this.worldPosition.getX();
         int j = this.worldPosition.getY();
         int k = this.worldPosition.getZ();
-        float f = 0.1F;
+        double d0;
+        // Paper end
+
         if (this.openCount > 0 && this.openness == 0.0F) {
             double d = (double)i + 0.5D;
             double e = (double)k + 0.5D;
             this.level.playSound((Player)null, d, (double)j + 0.5D, e, SoundEvents.ENDER_CHEST_OPEN, SoundSource.BLOCKS, 0.5F, this.level.random.nextFloat() * 0.1F + 0.9F);
         }
+        // Paper start
+    }
 
-        if (this.openCount == 0 && this.openness > 0.0F || this.openCount > 0 && this.openness < 1.0F) {
-            float g = this.openness;
-            if (this.openCount > 0) {
-                this.openness += 0.1F;
+    private void doCloseLogic() {
+        int i = this.worldPosition.getX();
+        int j = this.worldPosition.getY();
+        int k = this.worldPosition.getZ();
+        double d0;
+
+        if (this.openCount == 0) { /* && this.a > 0.0F || this.c > 0 && this.a < 1.0F) {
+        // Paper end
+            float f1 = this.a;
+
+            if (this.c > 0) {
+                this.a += 0.1F;
             } else {
-                this.openness -= 0.1F;
+                this.a -= 0.1F;
             }
 
-            if (this.openness > 1.0F) {
-                this.openness = 1.0F;
+            if (this.a > 1.0F) {
+                this.a = 1.0F;
             }
 
-            float h = 0.5F;
-            if (this.openness < 0.5F && g >= 0.5F) {
-                double l = (double)i + 0.5D;
-                double m = (double)k + 0.5D;
-                this.level.playSound((Player)null, l, (double)j + 0.5D, m, SoundEvents.ENDER_CHEST_CLOSE, SoundSource.BLOCKS, 0.5F, this.level.random.nextFloat() * 0.1F + 0.9F);
-            }
+            float f2 = 0.5F;
+
+            if (this.a < 0.5F && f1 >= 0.5F) {
+            // Paper start
+            */
+                d0 = (double) i + 0.5D;
+                double d2 = (double) k + 0.5D;
+
+            MCUtil.scheduleTask(10, () -> {
+                this.level.playSound((Player) null, d0, (double) j + 0.5D, d2, SoundEvents.ENDER_CHEST_CLOSE, SoundSource.BLOCKS, 0.5F, this.level.random.nextFloat() * 0.1F + 0.9F);
+            }, "Chest Sounds");
 
             if (this.openness < 0.0F) {
                 this.openness = 0.0F;
@@ -77,11 +106,13 @@ public class EnderChestBlockEntity extends BlockEntity implements TickableBlockE
     public void startOpen() {
         ++this.openCount;
         this.level.blockEvent(this.worldPosition, Blocks.ENDER_CHEST, 1, this.openCount);
+        doOpenLogic(); // Paper
     }
 
     public void stopOpen() {
         --this.openCount;
         this.level.blockEvent(this.worldPosition, Blocks.ENDER_CHEST, 1, this.openCount);
+        doCloseLogic(); // Paper
     }
 
     public boolean stillValid(Player player) {
diff --git a/src/main/java/net/minecraft/world/level/block/state/StateHolder.java b/src/main/java/net/minecraft/world/level/block/state/StateHolder.java
index 0811bbeb5a1dfd41825b92c6b12fe7aef52d2e66..222cea29c78ca06091366a292a5324dcb682d9e6 100644
--- a/src/main/java/net/minecraft/world/level/block/state/StateHolder.java
+++ b/src/main/java/net/minecraft/world/level/block/state/StateHolder.java
@@ -82,6 +82,7 @@ public abstract class StateHolder<O, S> {
         return Collections.unmodifiableCollection(this.values.keySet());
     }
 
+    public <T extends Comparable<T>> boolean contains(Property<T> iblockstate) { return this.hasProperty(iblockstate); } // Paper - OBFHELPER
     public <T extends Comparable<T>> boolean hasProperty(Property<T> property) {
         return this.values.containsKey(property);
     }
