From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joseph Hirschfeld <joe@ibj.io>
Date: Thu, 3 Mar 2016 02:39:54 -0600
Subject: [PATCH] Change implementation of (tile)entity removal list

use sets for faster removal

diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 422e583eeb1642ec7fb2f13a8a38d3bb844943de..77a7a4cc98e71b4422993f1401c070bdf188bbaa 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -1120,7 +1120,7 @@ public class ServerLevel extends net.minecraft.world.level.Level implements Worl
             }
         }
         // Spigot End
-        this.blockEntitiesToUnload.addAll(chunk.getBlockEntities().values());
+        this.tileEntityListUnload.addAll(chunk.getBlockEntities().values());
         List[] aentityslice = chunk.getEntitySlices(); // Spigot
         int i = aentityslice.length;
 
diff --git a/src/main/java/net/minecraft/world/level/Level.java b/src/main/java/net/minecraft/world/level/Level.java
index 14b456da5ed56028038000745aca6a719a23bb1c..717ceff6f4b03202aa76b4c49f21d64420a56de8 100644
--- a/src/main/java/net/minecraft/world/level/Level.java
+++ b/src/main/java/net/minecraft/world/level/Level.java
@@ -90,7 +90,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable {
     public final List<BlockEntity> blockEntityList = Lists.newArrayList();
     public final List<BlockEntity> tickableBlockEntities = Lists.newArrayList();
     protected final List<BlockEntity> pendingBlockEntities = Lists.newArrayList();
-    protected final List<BlockEntity> blockEntitiesToUnload = Lists.newArrayList();
+    protected final java.util.Set<BlockEntity> tileEntityListUnload = com.google.common.collect.Sets.newHashSet();
     public final Thread thread;
     private final boolean isDebug;
     private int skyDarken;
@@ -698,10 +698,10 @@ public abstract class Level implements LevelAccessor, AutoCloseable {
 
         gameprofilerfiller.push("blockEntities");
         timings.tileEntityTick.startTiming(); // Spigot
-        if (!this.blockEntitiesToUnload.isEmpty()) {
-            this.tickableBlockEntities.removeAll(this.blockEntitiesToUnload);
-            this.blockEntityList.removeAll(this.blockEntitiesToUnload);
-            this.blockEntitiesToUnload.clear();
+        if (!this.tileEntityListUnload.isEmpty()) {
+            this.tickableBlockEntities.removeAll(this.tileEntityListUnload);
+            this.blockEntityList.removeAll(this.tileEntityListUnload);
+            this.tileEntityListUnload.clear();
         }
 
         this.updatingBlockEntities = true;
