From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 28 Mar 2016 20:32:58 -0400
Subject: [PATCH] Entity AddTo/RemoveFrom World Events


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index de527f503822b85b560c66ea9afedd8b5dbb914a..b7dd76f3bc144dd4841f5d905daa60acfd294b4e 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -1209,7 +1209,7 @@ public class ServerLevel extends net.minecraft.world.level.Level implements Worl
         if (entity instanceof Mob) {
             this.navigations.remove(((Mob) entity).getNavigation());
         }
-
+        new com.destroystokyo.paper.event.entity.EntityRemoveFromWorldEvent(entity.getBukkitEntity()).callEvent(); // Paper - fire while valid
         entity.valid = false; // CraftBukkit
     }
 
@@ -1247,6 +1247,7 @@ public class ServerLevel extends net.minecraft.world.level.Level implements Worl
                 entity.origin = entity.getBukkitEntity().getLocation();
             }
             // Paper end
+            new com.destroystokyo.paper.event.entity.EntityAddToWorldEvent(entity.getBukkitEntity()).callEvent(); // Paper - fire while valid
         }
 
     }
