From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sat, 2 Apr 2016 05:09:16 -0400
Subject: [PATCH] Add PlayerUseUnknownEntityEvent


diff --git a/src/main/java/net/minecraft/network/protocol/game/ServerboundInteractPacket.java b/src/main/java/net/minecraft/network/protocol/game/ServerboundInteractPacket.java
index 6f5dea4d031763a261be1e74c079474e71f6593a..3d782275ee86ecbbe9ed0c3a2faf2ef2eff6e23b 100644
--- a/src/main/java/net/minecraft/network/protocol/game/ServerboundInteractPacket.java
+++ b/src/main/java/net/minecraft/network/protocol/game/ServerboundInteractPacket.java
@@ -10,7 +10,7 @@ import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.Vec3;
 
 public class ServerboundInteractPacket implements Packet<ServerGamePacketListener> {
-    private int entityId;
+    private int entityId; public int getEntityId() { return this.entityId; } // Paper - add accessor
     private ServerboundInteractPacket.Action action;
     private Vec3 location;
     private InteractionHand hand;
diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index da1d471b4f4c19f07730172f98ab5d29d3da9c90..600bc268b5ef67af69e156284f514f289e4bccf6 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2184,6 +2184,16 @@ public class ServerGamePacketListenerImpl implements ServerGamePacketListener {
                 }
             }
         }
+        // Paper start - fire event
+        else {
+            this.craftServer.getPluginManager().callEvent(new com.destroystokyo.paper.event.player.PlayerUseUnknownEntityEvent(
+                this.getPlayer(),
+                packetplayinuseentity.getEntityId(),
+                packetplayinuseentity.getAction() == ServerboundInteractPacket.Action.ATTACK,
+                packetplayinuseentity.getHand() == InteractionHand.MAIN_HAND ? EquipmentSlot.HAND : EquipmentSlot.OFF_HAND
+            ));
+        }
+        // Paper end
 
     }
 
