From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sat, 7 May 2016 23:33:08 -0400
Subject: [PATCH] Don't save empty scoreboard teams to scoreboard.dat


diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
index 750508267982046d0c08cffdd57c28459f7ce9c1..71cbf3a321dcf2ff2071d46ff2f117f512997f7d 100644
--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
@@ -229,4 +229,9 @@ public class PaperConfig {
     private static void enablePlayerCollisions() {
         enablePlayerCollisions = getBoolean("settings.enable-player-collisions", true);
     }
+
+    public static boolean saveEmptyScoreboardTeams = false;
+    private static void saveEmptyScoreboardTeams() {
+        saveEmptyScoreboardTeams = getBoolean("settings.save-empty-scoreboard-teams", false);
+    }
 }
diff --git a/src/main/java/net/minecraft/world/scores/ScoreboardSaveData.java b/src/main/java/net/minecraft/world/scores/ScoreboardSaveData.java
index 6db23cf2e8d0bb361faeed976eea9643c13802ab..af9bf3be7de13e2825a74cbf31b210f5aad7bc44 100644
--- a/src/main/java/net/minecraft/world/scores/ScoreboardSaveData.java
+++ b/src/main/java/net/minecraft/world/scores/ScoreboardSaveData.java
@@ -164,6 +164,7 @@ public class ScoreboardSaveData extends SavedData {
         ListTag listTag = new ListTag();
 
         for(PlayerTeam playerTeam : this.scoreboard.getPlayerTeams()) {
+            if (!com.destroystokyo.paper.PaperConfig.saveEmptyScoreboardTeams && playerTeam.getPlayers().isEmpty()) continue; // Paper
             CompoundTag compoundTag = new CompoundTag();
             compoundTag.putString("Name", playerTeam.getName());
             compoundTag.putString("DisplayName", Component.Serializer.toJson(playerTeam.getDisplayName()));
