From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Fri, 16 Dec 2016 16:03:19 -0600
Subject: [PATCH] Don't let fishinghooks use portals


diff --git a/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java b/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
index f79de132f69cd2ce100ce9d8d10808adbfa49db0..ce21756ab5e765a017db03b88207df2bc013b4d5 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
@@ -238,6 +238,11 @@ public class FishingHook extends Projectile {
 
             this.setDeltaMovement(this.getDeltaMovement().scale(0.92D));
             this.reapplyPosition();
+            // Paper start - These shouldn't be going through portals
+            if (this.isInsidePortal) {
+                this.remove();
+            }
+            // Paper end
         }
     }
 
