From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 23 Jul 2018 22:18:31 -0400
Subject: [PATCH] Mark chunk dirty anytime entities change to guarantee it
 saves


diff --git a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
index 59da7bea726dfc1e61ac3a7cb22e24d90014f2de..db25b14a64820c24ecc272b06a4c400c533a23d1 100644
--- a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
@@ -554,6 +554,7 @@ public class LevelChunk implements ChunkAccess {
         entity.zChunk = this.chunkPos.z;
         this.entities.add(entity); // Paper - per chunk entity list
         this.entitySlices[k].add(entity);
+        this.markUnsaved(); // Paper
     }
 
     @Override
@@ -579,6 +580,7 @@ public class LevelChunk implements ChunkAccess {
         if (!this.entitySlices[section].remove(entity)) {
             return;
         }
+        this.markUnsaved(); // Paper
         // Paper end
         this.entities.remove(entity); // Paper
     }
