From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Fri, 12 Oct 2018 01:37:22 -0500
Subject: [PATCH] Here's Johnny!


diff --git a/src/main/java/net/minecraft/world/entity/monster/Vindicator.java b/src/main/java/net/minecraft/world/entity/monster/Vindicator.java
index c728f023d3f496959d9af065c06efa9ace442071..1d0a4b94c340ac4a04d12fbc09d3bc8f295e0d78 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Vindicator.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Vindicator.java
@@ -50,7 +50,7 @@ public class Vindicator extends AbstractIllager {
     private static final Predicate<Difficulty> DOOR_BREAKING_PREDICATE = (difficulty) -> {
         return difficulty == Difficulty.NORMAL || difficulty == Difficulty.HARD;
     };
-    private boolean isJohnny;
+    private boolean isJohnny; public boolean isJohnny() { return isJohnny; } public void setJohnny(boolean johnny) { isJohnny = johnny; } // Paper - OBFHELPER
 
     public Vindicator(EntityType<? extends Vindicator> entityType, Level level) {
         super(entityType, level);
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftVindicator.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftVindicator.java
index 3c5994e9862e5caa257ee6a21f8fba2df39c98c5..6c1569340317f7bed39eaf6e858d602234993eb3 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftVindicator.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftVindicator.java
@@ -24,4 +24,14 @@ public class CraftVindicator extends CraftIllager implements Vindicator {
     public EntityType getType() {
         return EntityType.VINDICATOR;
     }
+
+    // Paper start
+    public boolean isJohnny() {
+        return getHandle().isJohnny();
+    }
+
+    public void setJohnny(boolean johnny) {
+        getHandle().setJohnny(johnny);
+    }
+    // Paper end
 }
