From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GreenMeanie <GreenMeanieMC@gmail.com>
Date: Sat, 20 Oct 2018 22:34:02 -0400
Subject: [PATCH] Reset players airTicks on respawn


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 9aad839bed6fde51b9eb79da1a3645efd3ec2f56..0ea1d58a40c8aed8e2856f91b1cc295230b40475 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -2130,6 +2130,7 @@ public class ServerPlayer extends Player implements ContainerListener {
         }
 
         this.setHealth(this.getMaxHealth());
+        this.setAirSupply(this.getMaxAirTicks()); // Paper
         this.remainingFireTicks = 0;
         this.fallDistance = 0;
         this.foodData = new FoodData(this);
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 3a2dfc32e4bd17d5e07e5ed804ac37d6bb463e99..bab518440ac283f587e414d0d3bb0916338128d2 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -2354,6 +2354,7 @@ public abstract class Entity implements Nameable, CommandSource, KeyedObject { /
 
     }
 
+    public final int getMaxAirTicks() { return getMaxAirSupply(); } // Paper - OBFHELPER
     public int getMaxAirSupply() {
         return 300;
     }
