From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sat, 2 May 2020 03:09:46 -0400
Subject: [PATCH] Validate PickItem Packet and kick for invalid


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 4fc0ac8ec7facf2b3fdaa4a3dbdaffc472b1bd8a..90e54183d3bb8a85f5a71bb51edd6b495355859a 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -877,7 +877,14 @@ public class ServerGamePacketListenerImpl implements ServerGamePacketListener {
     @Override
     public void handlePickItem(ServerboundPickItemPacket packetplayinpickitem) {
         PacketUtils.ensureRunningOnSameThread(packetplayinpickitem, this, this.player.getLevel());
-        this.player.inventory.pickSlot(packetplayinpickitem.getSlot());
+        // Paper start - validate pick item position
+        if (!(packetplayinpickitem.getSlot() >= 0 && packetplayinpickitem.getSlot() < this.player.inventory.items.size())) {
+            ServerGamePacketListenerImpl.LOGGER.warn("{} tried to set an invalid carried item", this.player.getName().getString());
+            this.disconnect("Invalid hotbar selection (Hacking?)");
+            return;
+        }
+        this.player.inventory.pickSlot(packetplayinpickitem.getSlot()); // Paper - Diff above if changed
+        // Paper end
         this.player.connection.send(new ClientboundContainerSetSlotPacket(-2, this.player.inventory.selected, this.player.inventory.getItem(this.player.inventory.selected)));
         this.player.connection.send(new ClientboundContainerSetSlotPacket(-2, packetplayinpickitem.getSlot(), this.player.inventory.getItem(packetplayinpickitem.getSlot())));
         this.player.connection.send(new ClientboundSetCarriedItemPacket(this.player.inventory.selected));
