From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 10 May 2020 22:12:46 -0400
Subject: [PATCH] Ensure Entity AABB's are never invalid


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index db673e00bd8729db0198b1e7f25dbea607345deb..067dffc4408f188dece4e6c9b7705b82ca6dc810 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -67,6 +67,7 @@ import net.minecraft.world.Nameable;
 import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.entity.animal.AbstractFish;
 import net.minecraft.world.entity.animal.Animal;
+import net.minecraft.world.entity.decoration.HangingEntity;
 import net.minecraft.world.entity.item.ItemEntity;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.entity.vehicle.AbstractMinecart;
@@ -481,7 +482,7 @@ public abstract class Entity implements Nameable, CommandSource, KeyedObject { /
 
     public void setPos(double x, double y, double z) {
         this.setPosRaw(x, y, z);
-        this.setBoundingBox(this.dimensions.makeBoundingBox(x, y, z));
+        //this.a(this.size.a(d0, d1, d2)); // Paper - move into setPositionRaw
         if (valid) ((ServerLevel) level).updateChunkPos(this); // CraftBukkit
     }
 
@@ -3425,6 +3426,12 @@ public abstract class Entity implements Nameable, CommandSource, KeyedObject { /
     }
 
     public void setPosRaw(double x, double y, double z) {
+        // Paper start - never allow AABB to become desynced from position
+        // hanging has its own special logic
+        if (!(this instanceof HangingEntity) && (this.position.x != x || this.position.y != y || this.position.z != z)) {
+            this.setBoundingBox(this.dimensions.makeBoundingBox(x, y, z));
+        }
+        // Paper end
         if (this.position.x != x || this.position.y != y || this.position.z != z) {
             this.position = new Vec3(x, y, z);
             int i = Mth.floor(x);
