From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: JRoy <joshroy126@gmail.com>
Date: Wed, 1 Jul 2020 18:01:49 -0400
Subject: [PATCH] Remove streams from classes related villager gossip


diff --git a/src/main/java/net/minecraft/world/entity/ai/gossip/GossipContainer.java b/src/main/java/net/minecraft/world/entity/ai/gossip/GossipContainer.java
index eadb7b4aaa98f1a910962a65445b94d5f068dea6..b9583904bc7e48fb291cbd34ab28318b0ef68a98 100644
--- a/src/main/java/net/minecraft/world/entity/ai/gossip/GossipContainer.java
+++ b/src/main/java/net/minecraft/world/entity/ai/gossip/GossipContainer.java
@@ -8,6 +8,7 @@ import com.mojang.serialization.Dynamic;
 import com.mojang.serialization.DynamicOps;
 import it.unimi.dsi.fastutil.objects.Object2IntMap;
 import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
+import it.unimi.dsi.fastutil.objects.ObjectArrayList; // Paper
 import it.unimi.dsi.fastutil.objects.ObjectIterator;
 import it.unimi.dsi.fastutil.objects.Object2IntMap.Entry;
 import java.util.Arrays;
@@ -47,8 +48,21 @@ public class GossipContainer {
         });
     }
 
+    // Paper start - Remove streams from reputation
+    private List<GossipContainer.GossipEntry> decompress() {
+        List<GossipContainer.GossipEntry> list = new ObjectArrayList<>();
+        for (Map.Entry<UUID, GossipContainer.EntityGossips> entry : getReputations().entrySet()) {
+            for (GossipContainer.GossipEntry cur : entry.getValue().decompress(entry.getKey())) {
+                if (cur.weightedValue() != 0)
+                    list.add(cur);
+            }
+        }
+        return list;
+    }
+    // Paper end
+
     private Collection<GossipContainer.GossipEntry> selectGossipsForTransfer(Random random, int count) {
-        List<GossipContainer.GossipEntry> list = this.unpack().collect(Collectors.toList());
+        List<GossipContainer.GossipEntry> list = decompress(); // Paper - Remove streams from reputation
         if (list.isEmpty()) {
             return Collections.emptyList();
         } else {
@@ -108,7 +122,7 @@ public class GossipContainer {
     }
 
     public <T> Dynamic<T> store(DynamicOps<T> dynamicOps) {
-        return new Dynamic<>(dynamicOps, dynamicOps.createList(this.unpack().map((gossipEntry) -> {
+        return new Dynamic<>(dynamicOps, dynamicOps.createList(this.decompress().stream().map((gossipEntry) -> { // Paper
             return gossipEntry.store(dynamicOps);
         }).map(Dynamic::getValue)));
     }
@@ -131,17 +145,29 @@ public class GossipContainer {
     }
 
     public static class EntityGossips { // Paper - make public
-        private final Object2IntMap<GossipType> entries = new Object2IntOpenHashMap<>();
+        private final Object2IntMap<GossipType> entries = new Object2IntOpenHashMap<>(); private Object2IntMap<GossipType> getEntries() { return entries; } // Paper - OBFHELPER
 
         public EntityGossips() { // Paper - make public - update CraftVillager setReputation on change
         }
 
         public int weightedValue(Predicate<GossipType> gossipTypeFilter) {
-            return this.entries.object2IntEntrySet().stream().filter((entry) -> {
-                return gossipTypeFilter.test(entry.getKey());
-            }).mapToInt((entry) -> {
-                return entry.getIntValue() * (entry.getKey()).weight;
-            }).sum();
+            // Paper start - Remove streams from reputation
+            int weight = 0;
+            for (Object2IntMap.Entry<GossipType> entry : getEntries().object2IntEntrySet()) {
+                if (gossipTypeFilter.test(entry.getKey())) {
+                    weight += entry.getIntValue() * entry.getKey().getWeight();
+                }
+            }
+            return weight;
+        }
+
+        public List<GossipContainer.GossipEntry> decompress(UUID uuid) {
+            List<GossipContainer.GossipEntry> list = new ObjectArrayList<>();
+            for (Object2IntMap.Entry<GossipType> entry : getEntries().object2IntEntrySet()) {
+                list.add(new GossipContainer.GossipEntry(uuid, entry.getKey(), entry.getIntValue()));
+            }
+            return list;
+            // Paper - end
         }
 
         public Stream<GossipContainer.GossipEntry> unpack(UUID target) {
diff --git a/src/main/java/net/minecraft/world/entity/ai/gossip/GossipType.java b/src/main/java/net/minecraft/world/entity/ai/gossip/GossipType.java
index d92c9424838ab50bfa72caf7116f32dac8e2ec0e..2f9c19daaed255abf59cf62c0e3e2b9ea18b0d23 100644
--- a/src/main/java/net/minecraft/world/entity/ai/gossip/GossipType.java
+++ b/src/main/java/net/minecraft/world/entity/ai/gossip/GossipType.java
@@ -14,7 +14,7 @@ public enum GossipType {
     TRADING("trading", 1, 25, 2, 20);
 
     public final String id;
-    public final int weight;
+    public final int weight; public int getWeight() { return weight; } // Paper - OBFHELPER
     public final int max;
     public final int decayPerDay;
     public final int decayPerTransfer;
