From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Thu, 2 Jul 2020 18:11:43 -0500
Subject: [PATCH] Add entity liquid API


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index a20eefa08d239556bfd47bf15da2fa72d188cb58..aef860fe97b48d50c6c6cd1838b6b9238ec48d89 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -1167,13 +1167,13 @@ public abstract class Entity implements Nameable, CommandSource, KeyedObject { /
         return this.wasTouchingWater;
     }
 
-    private boolean isInRain() {
+    public boolean isInRain() { // Paper - private -> public
         BlockPos blockposition = this.blockPosition();
 
         return this.level.isRainingAt(blockposition) || this.level.isRainingAt(new BlockPos((double) blockposition.getX(), this.getBoundingBox().maxY, (double) blockposition.getZ()));
     }
 
-    private boolean isInBubbleColumn() {
+    public boolean isInBubbleColumn() { // Paper - public
         return this.level.getBlockState(this.blockPosition()).is(Blocks.BUBBLE_COLUMN);
     }
 
@@ -1186,6 +1186,7 @@ public abstract class Entity implements Nameable, CommandSource, KeyedObject { /
         return this.isInWater() || this.isInRain() || this.isInBubbleColumn();
     }
 
+    public final boolean isInWaterOrBubbleColumn() { return isInWaterOrBubble(); } // Paper - OBFHELPER
     public boolean isInWaterOrBubble() {
         return this.isInWater() || this.isInBubbleColumn();
     }
