From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 5 Jul 2020 14:59:31 -0400
Subject: [PATCH] Don't check chunk for portal on world gen entity add


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 24b4d0071940082922a302681089fb59271eca02..33647a46ce4626b341c2a3e9f50d4bc78be0c9cc 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -3017,7 +3017,7 @@ public abstract class LivingEntity extends Entity {
         Entity entity = this.getVehicle();
 
         super.stopRiding(suppressCancellation); // Paper - suppress
-        if (entity != null && entity != this.getVehicle() && !this.level.isClientSide) {
+        if (entity != null && entity != this.getVehicle() && !this.level.isClientSide && entity.valid) { // Paper - don't process on world gen
             this.dismountVehicle(entity);
         }
 
