From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Andrew Steinborn <git@steinborn.me>
Date: Sun, 5 Jul 2020 22:38:18 -0400
Subject: [PATCH] Optimize NetworkManager Exception Handling


diff --git a/src/main/java/net/minecraft/network/ConnectionProtocol.java b/src/main/java/net/minecraft/network/ConnectionProtocol.java
index aecfa3e28fb0052c2d09c4ea030f7bdcadb38515..3200fad97eeeab611c04775980f7eccccc104b74 100644
--- a/src/main/java/net/minecraft/network/ConnectionProtocol.java
+++ b/src/main/java/net/minecraft/network/ConnectionProtocol.java
@@ -262,6 +262,7 @@ public enum ConnectionProtocol {
 
         @Nullable
         public Packet<?> createPacket(int id) {
+            if (id < 0 || id >= this.idToConstructor.size()) return null; // Paper
             Supplier<? extends Packet<T>> supplier = this.idToConstructor.get(id);
             return supplier != null ? supplier.get() : null;
         }
diff --git a/src/main/java/net/minecraft/network/Varint21FrameDecoder.java b/src/main/java/net/minecraft/network/Varint21FrameDecoder.java
index 17a85f14e92ef17c9a60a64e339e3ae4978a117a..c5a6ba4064f26e01796b096cc0621af403584aee 100644
--- a/src/main/java/net/minecraft/network/Varint21FrameDecoder.java
+++ b/src/main/java/net/minecraft/network/Varint21FrameDecoder.java
@@ -8,10 +8,20 @@ import io.netty.handler.codec.CorruptedFrameException;
 import java.util.List;
 
 public class Varint21FrameDecoder extends ByteToMessageDecoder {
+    private final byte[] lenBuf = new byte[3]; // Paper
     @Override
     protected void decode(ChannelHandlerContext channelHandlerContext, ByteBuf byteBuf, List<Object> list) throws Exception {
+        // Paper start - if channel is not active just discard the packet
+        if (!channelHandlerContext.channel().isActive()) {
+            byteBuf.skipBytes(byteBuf.readableBytes());
+            return;
+        }
+        // Paper end
         byteBuf.markReaderIndex();
-        byte[] bs = new byte[3];
+        // Paper start - reuse temporary length buffer
+        byte[] bs = lenBuf;
+        java.util.Arrays.fill(bs, (byte) 0);
+        // Paper end
 
         for(int i = 0; i < bs.length; ++i) {
             if (!byteBuf.isReadable()) {
diff --git a/src/main/java/net/minecraft/network/protocol/PacketUtils.java b/src/main/java/net/minecraft/network/protocol/PacketUtils.java
index cefa4b0696789fa7e23c19015a1212bf68a7d303..6f53729b74a0a92233456c8bd1d432e2befa3661 100644
--- a/src/main/java/net/minecraft/network/protocol/PacketUtils.java
+++ b/src/main/java/net/minecraft/network/protocol/PacketUtils.java
@@ -4,7 +4,10 @@ import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import co.aikar.timings.MinecraftTimings; // Paper
 import co.aikar.timings.Timing; // Paper
+import net.minecraft.network.Connection;
 import net.minecraft.network.PacketListener;
+import net.minecraft.network.chat.TextComponent;
+import net.minecraft.network.protocol.game.ClientboundDisconnectPacket;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.RunningOnDifferentThreadException;
 import net.minecraft.server.level.ServerLevel;
@@ -28,6 +31,21 @@ public class PacketUtils {
                     try (Timing ignored = timing.startTiming()) { // Paper - timings
                     packet.handle(listener);
                     } // Paper - timings
+                    // Paper start
+                    catch (Exception e) {
+                        Connection networkmanager = listener.getConnection();
+                        if (networkmanager.getPlayer() != null) {
+                            LOGGER.error("Error whilst processing packet {} for {}[{}]", packet, networkmanager.getPlayer().getScoreboardName(), networkmanager.getRemoteAddress(), e);
+                        } else {
+                            LOGGER.error("Error whilst processing packet {} for connection from {}", packet, networkmanager.getRemoteAddress(), e);
+                        }
+                        TextComponent error = new TextComponent("Packet processing error");
+                        networkmanager.send(new ClientboundDisconnectPacket(error), (future) -> {
+                            networkmanager.disconnect(error);
+                        });
+                        networkmanager.setReadOnly();
+                    }
+                    // Paper end
                 } else {
                     PacketUtils.LOGGER.debug("Ignoring packet due to disconnection: " + packet);
                 }
