From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Sat, 15 Aug 2020 08:04:49 -0500
Subject: [PATCH] Fix MC-187716 Use configured height


diff --git a/src/main/java/net/minecraft/world/level/levelgen/surfacebuilders/NetherCappedSurfaceBuilder.java b/src/main/java/net/minecraft/world/level/levelgen/surfacebuilders/NetherCappedSurfaceBuilder.java
index 00c2da17e822dcbe6c7c65956ecb4a6cb7b8aae4..a8b813ad83941b7a6b954abd7ea89ff816711a0e 100644
--- a/src/main/java/net/minecraft/world/level/levelgen/surfacebuilders/NetherCappedSurfaceBuilder.java
+++ b/src/main/java/net/minecraft/world/level/levelgen/surfacebuilders/NetherCappedSurfaceBuilder.java
@@ -42,7 +42,7 @@ public abstract class NetherCappedSurfaceBuilder extends SurfaceBuilder<SurfaceB
         BlockPos.MutableBlockPos mutableBlockPos = new BlockPos.MutableBlockPos();
         BlockState blockState5 = chunkAccess.getBlockState(mutableBlockPos.set(o, 128, p));
 
-        for(int s = 127; s >= 0; --s) {
+        for(int s = k; s >= 0; --s) { // Paper - fix MC-187716 - use configured height
             mutableBlockPos.set(o, s, p);
             BlockState blockState6 = chunkAccess.getBlockState(mutableBlockPos);
             if (blockState5.is(blockState.getBlock()) && (blockState6.isAir() || blockState6 == blockState2)) {
diff --git a/src/main/java/net/minecraft/world/level/levelgen/surfacebuilders/NetherForestSurfaceBuilder.java b/src/main/java/net/minecraft/world/level/levelgen/surfacebuilders/NetherForestSurfaceBuilder.java
index 48c49a47ca7164e91d56a302757085a215cdf10b..016cdfd296d188265f8e9f40a929fea2615200e2 100644
--- a/src/main/java/net/minecraft/world/level/levelgen/surfacebuilders/NetherForestSurfaceBuilder.java
+++ b/src/main/java/net/minecraft/world/level/levelgen/surfacebuilders/NetherForestSurfaceBuilder.java
@@ -34,7 +34,7 @@ public class NetherForestSurfaceBuilder extends SurfaceBuilder<SurfaceBuilderBas
         int r = -1;
         BlockState blockState3 = surfaceBuilderBaseConfiguration.getUnderMaterial();
 
-        for(int s = 127; s >= 0; --s) {
+        for(int s = k; s >= 0; --s) { // Paper - fix MC-187716 - use configured height
             mutableBlockPos.set(o, s, p);
             BlockState blockState4 = surfaceBuilderBaseConfiguration.getTopMaterial();
             BlockState blockState5 = chunkAccess.getBlockState(mutableBlockPos);
diff --git a/src/main/java/net/minecraft/world/level/levelgen/surfacebuilders/NetherSurfaceBuilder.java b/src/main/java/net/minecraft/world/level/levelgen/surfacebuilders/NetherSurfaceBuilder.java
index f3a2b12dc373e4469c2f70365e6db6c08cbe905c..464e0b7eca2bb9d2ed20bbbeb42f90669c1cd93a 100644
--- a/src/main/java/net/minecraft/world/level/levelgen/surfacebuilders/NetherSurfaceBuilder.java
+++ b/src/main/java/net/minecraft/world/level/levelgen/surfacebuilders/NetherSurfaceBuilder.java
@@ -36,7 +36,7 @@ public class NetherSurfaceBuilder extends SurfaceBuilder<SurfaceBuilderBaseConfi
         BlockState blockState3 = surfaceBuilderBaseConfiguration.getTopMaterial();
         BlockState blockState4 = surfaceBuilderBaseConfiguration.getUnderMaterial();
 
-        for(int s = 127; s >= 0; --s) {
+        for(int s = k; s >= 0; --s) { // Paper - fix MC-187716 - use configured height
             mutableBlockPos.set(o, s, p);
             BlockState blockState5 = chunkAccess.getBlockState(mutableBlockPos);
             if (blockState5.isAir()) {
