From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Sun, 23 Aug 2020 15:28:35 +0200
Subject: [PATCH] Add more Evoker API


diff --git a/src/main/java/net/minecraft/world/entity/monster/Evoker.java b/src/main/java/net/minecraft/world/entity/monster/Evoker.java
index 7455c0f598f80875de15373b49afd1c2d13506be..e0305b18de2d1c15318891e0333b3d782fc7214c 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Evoker.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Evoker.java
@@ -115,12 +115,12 @@ public class Evoker extends SpellcasterIllager {
         return SoundEvents.EVOKER_HURT;
     }
 
-    private void setWololoTarget(@Nullable Sheep sheep) {
+    public void setWololoTarget(@Nullable Sheep sheep) { // Paper - public
         this.wololoTarget = sheep;
     }
 
     @Nullable
-    private Sheep getWololoTarget() {
+    public Sheep getWololoTarget() { // Paper - public
         return this.wololoTarget;
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEvoker.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEvoker.java
index 29fec87c938c4252cf5c9473ce9e5c1908ea9063..49f94c30f9f639e5806c04ede91f1d468edf7eb1 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEvoker.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEvoker.java
@@ -35,4 +35,17 @@ public class CraftEvoker extends CraftSpellcaster implements Evoker {
     public void setCurrentSpell(Evoker.Spell spell) {
         getHandle().setIsCastingSpell(spell == null ? SpellcasterIllager.IllagerSpell.NONE : SpellcasterIllager.IllagerSpell.byId(spell.ordinal()));
     }
+
+    // Paper start
+    @Override
+    public org.bukkit.entity.Sheep getWololoTarget() {
+        net.minecraft.world.entity.animal.Sheep sheep = getHandle().getWololoTarget();
+        return sheep == null ? null : (org.bukkit.entity.Sheep) sheep.getBukkitEntity();
+    }
+
+    @Override
+    public void setWololoTarget(org.bukkit.entity.Sheep sheep) {
+        getHandle().setWololoTarget(sheep == null ? null : ((org.bukkit.craftbukkit.entity.CraftSheep) sheep).getHandle());
+    }
+    // Paper end
 }
