From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Wed, 18 Mar 2020 00:07:46 -0500
Subject: [PATCH] MC-147729: Drop items that are extra from a crafting recipe


diff --git a/src/main/java/net/minecraft/recipebook/ServerPlaceRecipe.java b/src/main/java/net/minecraft/recipebook/ServerPlaceRecipe.java
index e45f8bb0bbeb7bdbf3b4c75195f88e0c03b02a6c..58201b6e139e56aa7343718826d42b1e2dfce97c 100644
--- a/src/main/java/net/minecraft/recipebook/ServerPlaceRecipe.java
+++ b/src/main/java/net/minecraft/recipebook/ServerPlaceRecipe.java
@@ -67,7 +67,12 @@ public class ServerPlaceRecipe<C extends Container> implements PlaceRecipe<Integ
                 if (j == -1) {
                     j = this.inventory.getFreeSlot();
                 }
-
+                // Paper start
+                if (j == -1) {
+                    this.inventory.player.drop(itemStack.copy(), false);
+                    break;
+                }
+                // Paper end
                 ItemStack itemStack2 = itemStack.copy();
                 itemStack2.setCount(1);
                 if (!this.inventory.add(j, itemStack2)) {
