From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <spottedleaf@spottedleaf.dev>
Date: Sat, 25 Apr 2020 17:10:55 -0700
Subject: [PATCH] Reduce blockpos allocation from pathfinding


diff --git a/src/main/java/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java b/src/main/java/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
index 39108cd576eadbc8faed3c3831625d1f53df96aa..5fe62db14439a7b4314d9b123bbc684d1945c65b 100644
--- a/src/main/java/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
+++ b/src/main/java/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
@@ -463,7 +463,7 @@ public class WalkNodeEvaluator extends NodeEvaluator {
                             return BlockPathTypes.DANGER_FIRE;
                         }
 
-                        if (blockGetter.getFluidState(mutableBlockPos).is(FluidTags.WATER)) {
+                        if (iblockdata.getFluidState().is(FluidTags.WATER)) { // Paper - remove another getType call
                             return BlockPathTypes.WATER_BORDER;
                         }
                     }
@@ -491,7 +491,7 @@ public class WalkNodeEvaluator extends NodeEvaluator {
             } else if (blockState.is(Blocks.COCOA)) {
                 return BlockPathTypes.COCOA;
             } else {
-                FluidState fluidState = blockGetter.getFluidState(blockPos);
+                FluidState fluidState = blockState.getFluidState(); // Paper - remove another get type call;
                 if (fluidState.is(FluidTags.WATER)) {
                     return BlockPathTypes.WATER;
                 } else if (fluidState.is(FluidTags.LAVA)) {
