From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Anrza <andrzejrzeczycki314@gmail.com>
Date: Wed, 15 Jul 2020 12:08:49 +0200
Subject: [PATCH] Add LivingEntity#clearActiveItem


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index 2e464b82073ad079e61e6cda682f7a5b86066bf9..f133e57f2f6f15cbbed10713992902f9634e4627 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -763,6 +763,13 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         return getHandle().useItem.asBukkitMirror();
     }
 
+    // Paper start
+    @Override
+    public void clearActiveItem() {
+        getHandle().stopUsingItem();
+    }
+    // Paper end
+
     @Override
     public int getItemUseRemainingTime() {
         return getHandle().getItemUseRemainingTime();
