From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: KennyTV <jahnke.nassim@gmail.com>
Date: Tue, 25 Aug 2020 13:48:33 +0200
Subject: [PATCH] Add PlayerItemCooldownEvent


diff --git a/src/main/java/net/minecraft/world/item/ServerItemCooldowns.java b/src/main/java/net/minecraft/world/item/ServerItemCooldowns.java
index 3e47a7d29e561fbe4f2c90e5674ca8a00ff0bd20..ace08f8a24b79695550a31eb0684e1ce9f78a095 100644
--- a/src/main/java/net/minecraft/world/item/ServerItemCooldowns.java
+++ b/src/main/java/net/minecraft/world/item/ServerItemCooldowns.java
@@ -1,15 +1,26 @@
 package net.minecraft.world.item;
 
+import io.papermc.paper.event.player.PlayerItemCooldownEvent; // Paper
 import net.minecraft.network.protocol.game.ClientboundCooldownPacket;
 import net.minecraft.server.level.ServerPlayer;
 
 public class ServerItemCooldowns extends ItemCooldowns {
-    private final ServerPlayer player;
+    private final ServerPlayer player; public ServerPlayer getEntityPlayer() { return player; } // Paper - OBFHELPER
 
     public ServerItemCooldowns(ServerPlayer player) {
         this.player = player;
     }
 
+    // Paper start
+    @Override
+    public void addCooldown(Item item, int duration) {
+        PlayerItemCooldownEvent event = new PlayerItemCooldownEvent(getEntityPlayer().getBukkitEntity(), org.bukkit.craftbukkit.util.CraftMagicNumbers.getMaterial(item), duration);
+        if (event.callEvent()) {
+            super.addCooldown(item, event.getCooldown());
+        }
+    }
+    // Paper end
+
     protected void onCooldownStarted(Item item, int duration) {
         super.onCooldownStarted(item, duration);
         this.player.connection.send(new ClientboundCooldownPacket(item, duration));
